src/RootHitFiles.jl

```diff
@@ -0,0 +1,71 @@
+"""
+    RootHitFile(file::Union{IOStream, AbstractString})
+
+represents a `.root.hits` file, which can be iterated or read to yield events, containg data
+on hits.
+"""
+struct RootHitFile
+    stream::IOBuffer
+end
+
+RootHitFile(stream::IOStream) = RootHitFile(IOBuffer(mmap(stream)))
+RootHitFile(path::AbstractString) = occursin(r".root.hits$", path) ? RootHitFile((open(path))) :
+    throw(ArgumentError("$path is not a .root.hits file"))
+
+EventTuple = NamedTuple{
+    (:eventnum, :primcount, :pos, :E, :time, :particleID, :trkID, :trkparentID, :volumeID),
+    Tuple{
+        Int32, Int32, Vector{NTuple{3, Float32}}, Vector{Float32}, Vector{Float32},
+        Vector{Int32}, Vector{Int32}, Vector{Int32}, Vector{String}
+    }
+}
+
+function Base.read(f::RootHitFile)
+    eventnum  = Parsers.parse(Int32, f.stream)
+    hitcount  = Parsers.parse(Int32, f.stream)
+    primcount = Parsers.parse(Int32, f.stream)
+
+    # skip newline
+    skip(f.stream, 1)
+
+    pos         = Vector{NTuple{3, Float32}}(undef, hitcount)
+    E           = Vector{            Float32}(undef, hitcount)
+    time        = Vector{            Float32}(undef, hitcount)
+    particleID  = Vector{              Int32}(undef, hitcount)
+    trkID       = Vector{              Int32}(undef, hitcount)
+    trkparentID = Vector{              Int32}(undef, hitcount)
+    volumeID    = Vector{             String}(undef, hitcount)
+
+    @inbounds for i in 1:hitcount
+        pos[i] = (
+            Parsers.parse(Float32, f.stream),
+            Parsers.parse(Float32, f.stream),
+            Parsers.parse(Float32, f.stream)
+        )
+        E[i]           = Parsers.parse(Float32, f.stream)
+        time[i]        = Parsers.parse(Float32, f.stream)
+        particleID[i]  = Parsers.parse(Int32, f.stream)
+        trkID[i]       = Parsers.parse(Int32, f.stream)
+        trkparentID[i] = Parsers.parse(Int32, f.stream)
+        volumeID[i]    = String(readuntil(f.stream, UInt8('\n')))
+    end
+
+    return EventTuple((
+        eventnum, primcount, pos, E, time,
+        particleID, trkID, trkparentID, volumeID
+    ))
+end
+
+Base.eof(f::RootHitFile) = eof(f.stream)
+
+function Base.iterate(f::RootHitFile, state = nothing)
+    eof(f) && return nothing
+    return read(f), nothing
+end
+
+Base.IteratorSize(::Type{RootHitFile}) = Base.SizeUnknown()
+Base.IteratorEltype(::Type{RootHitFile}) = Base.HasEltype()
+Base.eltype(::RootHitFile) = EventTuple
+
+Tables.isrowtable(::Type{RootHitFile}) = true
+Tables.schema(f::RootHitFile) = Tables.Schema(eltype(f))
```

Comment:
Interesting approach, defining an on-disk table. Not sure about how far it will carry us here, because the file format (in contrast to HDF5) doesn't allow for random access. But let's try it. :-)

Reply:
We actually need out-of-memory capability because it allows us to run a lot more processes in parallel, thus significantly improving performance.

Reply:
And of course you can always do `collect(RootHitFile("path"))`, giving you the same thing. :)

Reply:
> We actually need out-of-memory capability

Yes, of course. I was just wondering if the Tables API is the right choice for streaming input, but let's give it a go find out in practice. Could you also add support for `Tables.partitions`, though, to allow block-wise processing? The blocks should be returned as `TypedTabes.Table` or `StructArrays.StructArray`. Block-size could be an option set by the user when opening the file. I have similar plans for our HDF5-I/O.

Reply:
I can add `Tables.partitions`, a global default batch size the user can change (like BenchmarkTools.jl does it) and a kw argument to the file to set it manually. That way it's just as easy to use and works similar to functions like pmap where it's also a kw argument.

I'm against returning any particular table though. I'd prefer returning an iterable of iterables and the user is free to collect it into whatever object they please. For those that want to use TypedTables or StructArrays they can construct them just like I would under the hood themselves. Importantly, for those for whom the raw iterable or some other table type is better, they lose nothing. If you want the iterable to be more eager, I'd be ok with `collect`ing the partitions, since that would return a vector of `NamedTuple`s. which is pretty much the default table.

Reply:
> global default batch size the user can change (like BenchmarkTools.jl does it) and a kw argument to the file to set it manually

Reply:
> global default batch size the user can change (like BenchmarkTools.jl does it) and a kw argument to the file to set it manually

A kwarg with a default should to nicely.

> I can add Tables.partitions [...] I'm against returning any particular table though.

Here, we do want to introduce certain standards across the LEGEND Julia ecosystem, to provide consistency to the users. It's perfectly fine if `RootHitFile` (you do, in fact, define a "particular table" here :-) ) is "just" an iterable over `EventTuple` (needs to be renamed, too generic) that implements the `Table` interface. However, `Tables.partitions` should return a an iterable over random access tables, and those should have a column-based memory layout. `StructArrays.StructArray` is the ideal choice here (and there's no alternative package if you want strucs instead of named tuples.).

Reply:
TLDR: I'm happy to do what you suggest. I was missing something in my thinking earlier.

Here's what I was missing: I can write the partitions into a column based format directly instead of writing a row table and then converting. I was always thinking of writing my tuples into a vector and then transforming it to a column-based format, which would be inefficient compared to just leaving it in the original partitioned format. That's why I was sceptical.

I'll have to add a little convenience method to allocate a NamedTuple of vectors of a given length. I guess this is why you suggested StructArrays, because they have this built in?

Reply:
> I was always thinking of writing my tuples into a vector and then transforming it to a column-based format, which would be inefficient

Reply:
> writing my tuples into a vector and then transforming it to a column-based format, which would be inefficient

Indeed, you'll want to build column-based storage directly for each partition, of course. However, in practice, the fact that we parse a non-column-based text file as input will probably be what limits the performance. :-)

Reply:
> I'll have to add a little convenience method to allocate a NamedTuple of vectors of a given length. I guess this is why you suggested StructArrays, because they have this built in?

No, StructArrays.jl is about storing arrays of generic strucs SoA-fashion, and TypesTables.jl is about storing arrays of `NamedTuple`s SoA-fashion (column-wise).

If you create an empty `TypedTables.Table` you can push into it. If you do know the number of rows, you can also pre-allocate the storage, of course.

To be efficient, though, you should use an `ArraysOfArrrays.VectorOfVectors` to back all columns that contains vectors. Othewise you're be allocation a separate vector for each vector-valued field in `RootHitEventTuple`, for each row. Since those vectors will usually be short, that would be very wasteful. `VectorOfVectors` stores a vector of variable-length vectors as a single contiguous vector under the hood. I use it for waveforms, too.

Btw: Longer term, we need to look at all the MC-inputs we have now and standardize column names and types (the ones they have in common, semantically).

Reply:
Oh yeah, `VectorOfVectors` is a good idea.

And yes, I was going to ask about the naming. My naming of columns is only partly based on what I saw legend packages using elsewhere, but we should figure out naming conventions for this stuff.

> If you create an empty TypedTables.Table you can push into it. If you do know the number of rows, you can also pre- allocate the storage, of course.

What I meant is that StructArrays supports `StructArray{MyStruct}(undef, 5)`. Does TypedTables have something like that? I didn't find it in the source. Either way, I'll have to make a custom method to be able to use `VectorOfVectors`.

Also, for the VolID field I might go to something like a categorical array, to avoid having a `Vector{String}`.

On the performance question, it's true that converting everything to a column based format does slow things down a bit. The main bottleneck is still parsing of text though. The old version of the code was fully row based, so I have the old benchmarks somewhere.

Reply:
> And yes, I was going to ask about the naming. My naming of columns

Just choose something for now. We'll iterate with @sagitta42 afterwards to harmonize things across input formats.

Reply:
> Does TypedTables have something like that?

Reply:
> What I meant is that StructArrays supports StructArray{MyStruct}(undef, 5). Does TypedTables have something like that?

No, you have to create the column vectors manually. But you'll have to do that anyhow to ensure you use column types like `VectorOfVectors` (it's semantic knowledge, neither TypedTables not StructArrays can guess that).

Reply:
> Also, for the VolID field I might go to something like a categorical array, to avoid having a Vector{String}

Yes, `String`s are out of the question. Is it a finite small set of values? Then we can use an enum. Otherwise we'll have to use a caterorical array (I think there are two or three different ones on the Julia ecosystem currently).

Reply:
> it's true that converting everything to a column based format does slow things down a bit. 

No matter, since at one point we'd want to go to columns anyhow.